### PR TITLE
KAFKA-12572: Add import ordering checkstyle rule and configure an automatic formatter

### PR DESCRIPTION
**Pull request by dongjinleekr**
This PR is a part of [KAFKA-10787](https://issues.apache.org/jira/browse/KAFKA-10787) umbrella task, replacing the previous implementation [here](https://github.com/apache/kafka/pull/8404/). **In short, this PR provides automatic import ordering and checking configurations, which will be a foundation of actual code formatting tasks later.** Here are some backgrounds.

When I first started [KAFKA-10787](https://issues.apache.org/jira/browse/KAFKA-10787), I thought it is a trivial one and can be resolved with only one PR. But it is entirely wrong.

1. It modifies hundreds of files; it is not only hard to review but also, every time it has some code conflict with the trunk branch.
2. As @mjsax already pointed out, the previous implementation omits the essential way of solving this problem - **automatic import formatter**. Since it includes the checkstyle rules only, it shows error messages for the import ordering violation and that's it.

For the reasons above, it is almost impossible to resolve this issue with one commit or PR.

So, I restarted this feature from scratch. First of all, I changed the approach like following:

1. Take an incremental approach, i.e., enlarge the import ordering rules applied area gradually. (In other words, narrow down the exception area to the import ordering rules, step by step.)
2. By adopting automatic code formatter, make the build tool to apply the import order automatically.

This approach has the advantages like:

1. We can apply the import order step by step; it is feasible to review and greatly reduces the possibility of code conflict.
2. The dev team doesn't have to care about the formatting anymore; the build tool automatically take cares of it.

With this scheme, we can expand the scope of auto-formatting and checkstyle incrementally.

In turn, I converted [KAFKA-10787](https://issues.apache.org/jira/browse/KAFKA-10787) into an umbrella one. As the first subtask, this PR do the following:

1. Define the import ordering rule in checkstyle: `kafka`, `org.apache.kafka`, `com`, `net`, `org`, `java`, `javax` and static imports. It is identical to [what we discussed in the dev mailing list](https://lists.apache.org/thread.html/rf6f49c845a3d48efe8a91916c8fbaddb76da17742eef06798fc5b24d%40%3Cdev.kafka.apache.org%3E), except each package is grouped independely. (e.g., `kafka` and `org.apache.kafka` are regarded as independent groups.)
2. Add an eclipse formatter file, which includes:
    - Minimal formatting rules that follow current checkstyle rules.
    - `ImportOrder` rule for the newly introduced import ordering rule.
3. **Add a spotless configuration to reformat the subset of modules automatically.**

Here are some reasons for the decisions I made:

1. Why choose eclipse formatter?

    As all of you know, Apache Kafka is using gradle as its build tool. To automatically apply the import ordering rule with gradle, we have to use [spotless](https://github.com/diffplug/spotless), which supports two formatters - google formatter and eclipse formatter.
    
    If we choose google formatter, it changes not only import ordering but also the other aspects of code, like an indentation. To minimize non-import ordering modifications, I choose eclipse formatter, which supports a custom formatter configuration.

    As you can see here, the formatter defines a minimal rules for out code only.

2. The original proposal consists of 3 groups: `kafka`/`org.apache.kafka`, `com`/`org`/`net`, `java`/`javax`, and static imports. Why are those packages separated into independent groups?

    It's simple - the spotless automatic formatter doesn't allow packages with different prefixes into one group. So I separated them.

**\* Updated (June 10th 2021): The `core` module is excluded from the PR, and it now only focuses on providing auto-formatting configurations.**

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dongjinleekr** [comment]
I think @ijuma would be the best reviewer for the new gradle configuration introduced here.

@hachikuji @mjsax @ableegoldman @cadonna Could you have a look? The automatic formatter formatted `core` module here. Have a try! :smiley:

**Comment by dongjinleekr** [comment]
@cadonna Hi Bruno, Congratulations on becoming a committer.

Could you have a look? :bow:

File: README.md
Comment (cadonna) [inline README.md]:
```suggestion
As of present, the auto-formatting configuration is work in progress. Auto-formatting is automatically invoked for the modules listed below when the 'checkstyleMain' or 'checkstyleTest' task is run.
```

File: eclipse-formatter.xml
Comment (cadonna) [inline eclipse-formatter.xml:1]: I guess we need to the license header in this file similarly as in `checkstyle.xml`.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (cadonna) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: I think that does not conform to our code style. I think it should look like:
```suggestion
    private Map<TopicPartition, Long> checkOffsetRange(final Map<TopicPartition, Long> inputTopicPartitionsAndOffset,
                                                       final Map<TopicPartition, Long> beginningOffsets,
                                                       final Map<TopicPartition, Long> endOffsets) {
```

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (cadonna) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Also this seems weird. I think this is because the line is 121 characters long.

File: eclipse-formatter.xml
Comment (cadonna) [inline eclipse-formatter.xml]: http://kafka.apache.org/coding-guide.html says:

"There is not a maximum line length (certainly not 80 characters, we don't work on punch cards any more), but be reasonable."

In Streams, we prefer a maximum line length of 120 but we do not enforce it with checkstyle (at least I could not find a rule for it). While I do not like long lines, I also think that sometimes there are good reasons to exceed the limit. So I would not enforce it in the formatter.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (cadonna) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: I guess the line break here was inserted on purpose. At least in Streams, we try to break after each sentence in javadocs. It would be odd if the formatter would remove the line break.

File: eclipse-formatter.xml
Comment (cadonna) [inline eclipse-formatter.xml:20]: Would it be possible to have only the import ordering in the formatter file and adding the other formatting rules incrementally?

**Review by cadonna — COMMENTED** [review]
@dongjinleekr Thank you for your perseverance with this!

I did a first pass. Here is my feedback.

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml:1]: Agree.

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml]: Agree. It seems like removing this rule would be better.

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml:20]: Let me have a try. 👍

**Comment by dongjinleekr** [comment]
Hi @cadonna,

Thanks again for your detailed review. Here is the update.

Following your suggestions, I tried some other configurations. First of all, there is no way to use automatic import ordering only - without any configuration, the formatter formats the code with its default configuration. So, we have to find the least-effecting configuration.

The major differences between the last draft are like the following:

- **Remove line wrapping rule for 120 characters.**
- **Comment is not formatted anymore.**
- Remove bracing-related rules.

But, I kept the other rules, like:

- As of present, both of 4 space indenting and 8 space indenting are co-existing in the code. I configured to use 4 space indenting for its majority.
- Several styles for the parameter formatting for the class and method definition also exist in the code. It is configured to the least-effecting configuration.

Please have a look when you are free. Thanks again for the review!

+1. It would be better to check how the new configuration formats streams module. I tested with that module, for it has so many `*.java` files. :wink:

File: eclipse-formatter.xml
Comment (cadonna) [reply eclipse-formatter.xml:1]: Do you still plan to add the license header here?

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (cadonna) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Although you removed the line length limit, the formatter still breaks here. Is 120 maybe the default of the eclipse formatter? Maybe it would make sense to set the limit to a high value (like 200) and decide case by case where to break as we actually do now.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (cadonna) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: We (Streams) usually break the arguments of a function call as follows:
```suggestion
        intermediateTopicsOption = optionParser.accepts(
            "intermediate-topics",
            "Comma-separated list of intermediate user topics (topics that are input and output topics, e.g., used in the deprecated through() method). For these topics, the tool will skip to the end."
        )
```

or 

```suggestion
        intermediateTopicsOption = optionParser.accepts(
            "intermediate-topics",
            "Comma-separated list of intermediate user topics (topics that are input and output topics, e.g., used in the deprecated through() method). For these topics, the tool will skip to the end.")
```
IMO both are better readable than having arguments filling up the line and then break. If all arguments fit in one line I would not break them. So my preferred formatting would be either all arguments fit in one line or each argument on its own line.

File: core/src/test/java/kafka/test/MockController.java
Comment (cadonna) [inline core/src/test/java/kafka/test/MockController.java]: IMO, this would be better readable as
```suggestion
    public CompletableFuture<Map<ConfigResource, ApiError>> incrementalAlterConfigs(Map<ConfigResource,    
                                                                                    Map<String, Map.Entry<AlterConfigOp.OpType, String>>> configChanges,
                                                                                    boolean validateOnly) {
```
(Note: GitHub suggestions make it actually less readable 😅. Each parameter should be on its own line (except the first) and be aligned after the `(` of the method.)

File: core/src/test/java/kafka/testkit/KafkaClusterTestKit.java
Comment (cadonna) [inline core/src/test/java/kafka/testkit/KafkaClusterTestKit.java]: I think breaking after `stream()` is on purpose here. At least, I use a similar style when I use Java's Stream API to make the chain of operators more readable. Would be a pity if the formatter would change it.

**Review by cadonna — COMMENTED** [review]
@dongjinleekr Thank you for the updates!

I left some comments. The comments about readability are quite subjective. Would be interesting to hear other opinions.

I would try to minimize changes of line breaks by the formatter, because I think most of the people add line breaks to make the code more readable. I have the impression that adding and removing line breaks is something that is hard to automate because it is often a case by case decision.

An exception to that are line breaks in method calls and method definitions. I think having a  formatter rule for that would be great.

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml:1]: OMG, I omitted it in the last update. :sweat:

**Comment by dongjinleekr** [comment]
Hi @cadonna, Here is the fix.

1. Added a license header to eclipse-formatter.xml.
2. Re-add line break formatting with 200 characters. (It resolved lots of formatting issues!)

+1. in `ClusterTestExtensions.java`, a space is added between `default` and `:`, resulting `default :`; It seems like a bug in eclipse formatter. :disappointed:

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk.

@cadonna Could wou have a look now? :smiley:

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk. cc/ @cadonna

File: eclipse-formatter.xml
Comment (cadonna) [inline eclipse-formatter.xml:55]: This line seems to equal to line 38.

File: eclipse-formatter.xml
Comment (cadonna) [inline eclipse-formatter.xml]: Personally, I would not enforce this.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java
Comment (cadonna) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: This contradicts
```
 <setting id="org.eclipse.jdt.core.formatter.insert_space_before_colon_in_case" value="do not insert"/>
```
in `eclipse-formatter.xml`.

**Review by cadonna — APPROVED** [review]
@dongjinleekr Thank you for the update!

I had just some minor comments.

LGTM! 

But I think also somebody that contributed to the core module should look at this PR. @ijuma @hachikuji @cmccabe @rajinisivaram @omkreddy @dajac

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml:55]: Agree. Fixed.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java
Comment (dongjinleekr) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Yes, I found that the eclipse formatter handles the default case independely. Additional configuration added to fix this. :+1:

File: eclipse-formatter.xml
Comment (dongjinleekr) [reply eclipse-formatter.xml]: Okay. Removed.

**Comment by dongjinleekr** [comment]
@cadonna I greatly appreciate your detailed review. Here is the update! :smile:

**Comment by dejan2609** [comment]
Shameless plug and related to CheckStyle:  ⏩  #10698 (needs a review / approval).

File: core/src/main/java/kafka/metrics/FilteringJmxReporter.java
Comment (cadonna) [inline core/src/main/java/kafka/metrics/FilteringJmxReporter.java]: @dongjinleekr The empty lines at the end of the class are still there although you removed the formatting directive. Maybe it is the default behavior. Could you try to add it again and set the value to 0?

**Comment by dongjinleekr** [comment]
@cadonna Here it is, with the following: 1. making server-common to ignore import orders. 2. rebasing onto the latest trunk.

+1. It seems like we can make the eclipse formatter do not insert a newline at the end of the class, but not removing it. (see [here](https://help.eclipse.org/2019-12/index.jsp?topic=%2Forg.eclipse.jdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjdt%2Fcore%2Fformatter%2FDefaultCodeFormatterConstants.html).) So I reformatted the `core` module from the latest trunk.

File: core/src/test/java/kafka/test/MockController.java
Comment (cadonna) [inline core/src/test/java/kafka/test/MockController.java]: One question here. If you put the parameters of this method on the same line as the method name and you run again the formatter, will the parameters stay on the same line as the method name?
What would happen if you put the first parameter on the same line as the method name and then put each of the remaining parameters on its line own line?

Comment (dongjinleekr) [reply core/src/test/java/kafka/test/MockController.java]: case 1. If you put the parameters of this method on the same line as the method name and run formatter: remains one line.
case 2. If you put the parameters separated and run formatter: broken. (below)

Comment (cadonna) [reply core/src/test/java/kafka/test/MockController.java]: I do not know what you exactly mean with "broken", but that is the behavior I hoped for.

Comment (dongjinleekr) [reply core/src/test/java/kafka/test/MockController.java]: Okay. I will fix it with the second approach.

Comment (cadonna) [reply core/src/test/java/kafka/test/MockController.java]: Could you put the first parameter in the same line as the method name and put the remaining parameters on their own line as I proposed? Could you do the same for `incrementalAlterConfigs()`?

If anybody does not like this style, we can still delete
```
<setting id="org.eclipse.jdt.core.formatter.alignment_for_parameters_in_constructor_declaration" value="18"/>
<setting id="org.eclipse.jdt.core.formatter.alignment_for_parameters_in_method_declaration" value="18"/>
```
from the formatter configuration, right?

Comment (cadonna) [reply core/src/test/java/kafka/test/MockController.java]: After that, I think, we can merge. Besides the import order changes, you kept the just a few formatting style changes, which I think is good. If somebody does not agree, we can make it even less invasive.

**Comment by dongjinleekr** [comment]
@cadonna Here it is. Rebased onto the latest trunk + manually formatted `MockController#{legacyAlterConfigs, incrementalAlterConfigs}`. I also reviewed the other files and it seems no more similar cases exist.

**Comment by dongjinleekr** [comment]
@cadonna Here is the update. I excluded the `core` module from this PR (i.e., it contains formatting configurations only, not actual formatting now) and opened separate issues:

- [KAFKA-12910: Configure automatic formatter for org.apache.kafka.streams.state](https://issues.apache.org/jira/browse/KAFKA-12910)
- [KAFKA-12911: Configure automatic formatter for org.apache.kafka.streams.processor](https://issues.apache.org/jira/browse/KAFKA-12911)

As soon as this PR is merged, I start the issues above.

File: build.gradle
Comment (cadonna) [inline build.gradle:647]: Don't you need to remove this to not format the code in core?

File: README.md
Comment (cadonna) [inline README.md]: I would remove this until we have it setup for Streams.

**Review by cadonna — COMMENTED** [review]
@dongjinleekr Thank you for the updates and your patience!

 I a few comments and I think you need to update the PR description.

File: build.gradle
Comment (dongjinleekr) [reply build.gradle:647]: Yes, we need this configuration; These dependencies are applied to the whole project, not core module only.

File: README.md
Comment (dongjinleekr) [reply README.md]: I thought it would be better to update the list every time the streams' packages are configured to auto-formatting - and since I am ready for this task, it will not need much time.

Moreover, These descriptions are applied to the whole project, isn't it?

Comment (cadonna) [reply README.md]: Yes, but before at least one module applies automatic formatting, I would remove this description because until then it is useless.

File: build.gradle
Comment (cadonna) [reply build.gradle:647]: Ah, now I see. I missed the if statement for the modules above on line 602.

File: README.md
Comment (dongjinleekr) [reply README.md]: Okay. I will move this section into the other subissue. :+1:

**Comment by dongjinleekr** [comment]
@cadonna Here it is. Rebased onto the latest trunk and removed the formatting section from `README.md`.

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk. cc/ @cadonna

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk. Also, I reflected the recent project structure change (i.e., `storage` → `storage/api`) by updating `suppressions.xml`. cc/ @cadonna @ableegoldman

**Comment by dongjinleekr** [comment]
Retest this please.

**Comment by chia7712** [comment]
@dongjinleekr Do you have free time to fix the conflicts? I'd like to review it!

**Comment by mimaison** [comment]
This was done in https://github.com/apache/kafka/commit/342e69192f62b89b7d1ea824aeecc09c38899d72, closing

**Pull request closed without merging**
